lib/mp4/captions-parser.js

```diff
@@ -0,0 +1,344 @@
+};
+
+/**
+  * Parses out caption nals from an FMP4 segment's video tracks.
+  *
+  * @param {Uint8Array} segment - The bytes of a single segment
+  * @param {Array<Number>} videoTrackIds - the trackIds of video tracks in the segment
+  * @return {Object.<Number, Object[]>} A mapping of video trackId to
+  *   a list of seiNals found in that track
+ **/
+var parseCaptionNals = function(segment, videoTrackIds) {
+  // To get the samples
+  var trafs = probe.findBox(segment, ['moof', 'traf']);
+  // To get SEI NAL units
+  var mdats = probe.findBox(segment, ['mdat']);
+  var captionNals = {};
+  var trafMdatPairs = [];
+
+  // Pair up each traf with a mdat
+  trafs.forEach(function(traf, index) {
+    var matchingMdat = mdats[index];
+    trafMdatPairs.push({
+      mdat: matchingMdat,
+      traf: traf
+    });
+  });
+
+  trafMdatPairs.forEach(function(pair) {
+    var mdat = pair.mdat;
+    var traf = pair.traf;
+    var tfhd = probe.findBox(traf, ['tfhd']);
+    // Exactly 1 tfhd per traf
+    var headerInfo = inspect.parseTfhd(tfhd[0]);
+    var trackId = headerInfo.trackId;
+    var tfdt = probe.findBox(traf, ['tfdt']);
+    // Either 0 or 1 tfdt per traf
+    var baseMediaDecodeTime = (tfdt.length > 0) ? inspect.parseTfdt(tfdt[0]).baseMediaDecodeTime : 0;
+    var truns = probe.findBox(traf, ['trun']);
+    var samples;
+    var seiNals;
+
+    // Only parse video data
+    if (videoTrackIds.includes(trackId) && truns.length > 0) {
+      samples = parseSamples(truns, baseMediaDecodeTime, headerInfo);
+
+      samples.sort(function(a, b) {
+        // This shouldn't ever happen
+        if (a.dataOffset === b.dataOffset) {
+          return a.pts - b.pts;
+        }
+        return a.dataOffset - b.dataOffset;
+      });
+
+      seiNals = findSeiNals(mdat, samples, trackId);
+
+      if (!captionNals[trackId]) {
+        captionNals[trackId] = [];
+      }
+
+      captionNals[trackId] = captionNals[trackId].concat(seiNals);
+    }
+  });
+
+  return captionNals;
+};
+
+/**
+  * Parses out inband captions from an MP4 container and returns
+  * caption objects that can be used by WebVTT and the TextTrack API.
+  * @see https://developer.mozilla.org/en-US/docs/Web/API/VTTCue
+  * @see https://developer.mozilla.org/en-US/docs/Web/API/TextTrack
+  *
+  * @param {Uint8Array} init - The init segment
+  * @param {Uint8Array} segment - The fmp4 segment containing imbedded captions
+  * @return {?Object[]} parsedCaptions - A list of captions or null if no video tracks
+  * @return {Number} parsedCaptions[].startTime - The time to show the caption in seconds
+  * @return {Number} parsedCaptions[].endTime - The time to stop showing the caption in seconds
+  * @return {String} parsedCaptions[].text - The visible content of the caption
+ **/
+var parseEmbeddedCaptions = function(init, segment) {
+  var captionStreams = {};
+  var parsedCaptions = [];
+
+  var videoTrackIds = getVideoTrackIds(init);
+  var captionStream = new CaptionStream();
```

Comment:
The `CaptionStream` needs to be persistent for the duration of the current media. It needs to have memory since the caption data can be spanned across several segments without being flushed.

IMO, it makes the most sense to make the object and call `push` and `reset` as appropriate directly from VHS, since it's what is doing the actual controlling. For example, right now for HLS, when the caption parser needs to be reset (which is only known by VHS), a `reset` signal is sent through media-sources to the transmuxer WebWorker. The reason that's so indirect is because VHS doesn't have a direct reference to the caption parser.

The transmuxer could potentially fire off an event when it parses NALs instead of what it's currently doing and passing it through a pipeline to a caption parser. That way the API from the VHS side would be mostly similar between HLS and DASH.

What do you think?

Reply:
I think we would want to avoid using the transmuxer for fmp4 segments as well since the transmuxer currently assumes it's going to see Annex B data. However, I'm working on restructuring this so VHS can `init` the CaptionsParser and then `parse` the segments it gets.

The approach I'm taking right now would still  parse the nals out, `push` nals into the CaptionStream and `flush` them from mux as, at least to me, it seems that VHS will always want to do that in the same place(once we have the segment downloaded).

Is there a scenario where you think we might want to `flush` the CaptionStream later? I think I can understand why we would want to reset the stream on rendition changes, e.c.t, from VHS.